src/obfuscation.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2025-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_OBFUSCATION_H
+#define BITCOIN_OBFUSCATION_H
+
+#include <span.h>
+#include <tinyformat.h>
+
+#include <array>
+#include <bit>
+#include <climits>
+#include <stdexcept>
+
+class Obfuscation
+{
+public:
+    static constexpr size_t SIZE_BYTES{sizeof(uint64_t)};
+
+    Obfuscation(const uint64_t key) { SetRotations(key); }
+    Obfuscation(const std::span<const std::byte, SIZE_BYTES> key_span) : Obfuscation(ToUint64(key_span)) {}
+
+    uint64_t Key() const { return m_rotations[0]; }
+    operator bool() const { return Key() != 0; }
+    void operator()(std::span<std::byte> target, const size_t key_offset_bytes = 0) const
+    {
+        if (!*this) return;
+        const uint64_t rot_key{m_rotations[key_offset_bytes % SIZE_BYTES]}; // Continue obfuscation from where we left off
+        for (; target.size() >= SIZE_BYTES; target = target.subspan(SIZE_BYTES)) { // Process multiple bytes at a time
+            Xor(target, rot_key, SIZE_BYTES);
+        }
```

Comment:
informational: Tried to optimize for aligned writes:
```suggestion
        uint64_t rot_key{m_rotations[key_offset_bytes % SIZE_BYTES]}; // Continue obfuscation from where we left off
        const size_t alignment_remaining{std::min(SIZE_BYTES - (reinterpret_cast<ptrdiff_t>(target.data()) % SIZE_BYTES), target.size())};
        Xor(target, rot_key, alignment_remaining);
        target = target.subspan(alignment_remaining);
        rot_key = m_rotations[(key_offset_bytes + alignment_remaining) % SIZE_BYTES];
        for (; target.size() >= SIZE_BYTES; target = target.subspan(SIZE_BYTES)) { // Process multiple bytes at a time
            *reinterpret_cast<uint64_t*>(target.data()) ^= rot_key;
        }
```
2x speedup on Clang for `ObfuscationBench`. But no difference, or a few percent worse on GCC.

Reply:
One more thing - while I was testing this out, I couldn't find a unit test that tested obfuscating a target buffer starting at different offsets in memory. Might be good to add one?

Reply:
`xor_bytes_reference` randomly generates where the next offset will be, check it by adding:
```C++
    void operator()(std::span<std::byte> target, const size_t key_offset_bytes = 0) const
    {
        if (!*this) return;
        const uint64_t rot_key{m_rotations[key_offset_bytes % SIZE_BYTES]}; // Continue obfuscation from where we left off
        std::cout << "key_offset_bytes % SIZE_BYTES = " << (key_offset_bytes % SIZE_BYTES) << std::endl;
```
which reveals:
```
key_offset_bytes % SIZE_BYTES = 4
key_offset_bytes % SIZE_BYTES = 0
key_offset_bytes % SIZE_BYTES = 7
key_offset_bytes % SIZE_BYTES = 4
key_offset_bytes % SIZE_BYTES = 0
key_offset_bytes % SIZE_BYTES = 1
key_offset_bytes % SIZE_BYTES = 6
key_offset_bytes % SIZE_BYTES = 1
key_offset_bytes % SIZE_BYTES = 7
key_offset_bytes % SIZE_BYTES = 0
key_offset_bytes % SIZE_BYTES = 3
key_offset_bytes % SIZE_BYTES = 3
key_offset_bytes % SIZE_BYTES = 0
key_offset_bytes % SIZE_BYTES = 5
key_offset_bytes % SIZE_BYTES = 2
...
```

---

I'll investigate the alignment and whether we can use SIMD in GCC as well (similarly to how Clang auto-detects it)

Reply:
I'm not sure about the alignment part, seems dangerous - do you see an explanation in the assembly why it would be faster?
I tried reproducing it, but I'm getting a significant slowdown instead (ran each 3x):

<details>
<summary>Current loop:</summary>

```C++
void operator()(std::span<std::byte> target, const size_t key_offset_bytes = 0) const
{
    if (!*this) return;
    const uint64_t rot_key{m_rotations[key_offset_bytes % SIZE_BYTES]}; // Continue obfuscation from where we left off
    for (; target.size() >= SIZE_BYTES; target = target.subspan(SIZE_BYTES)) { // Process multiple bytes at a time
        Xor(target, rot_key, SIZE_BYTES);
    }
    Xor(target, rot_key, target.size());
}
```

</details>

|              ns/MiB |               MiB/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           14,543.90 |           68,757.33 |    0.1% |     11.01 | `ObfuscationBench`
|           14,604.52 |           68,471.95 |    0.0% |     11.01 | `ObfuscationBench`
|           14,573.71 |           68,616.72 |    0.2% |     11.04 | `ObfuscationBench`

<details>
<summary>Suggestion:</summary>

```C++
void operator()(std::span<std::byte> target, const size_t key_offset_bytes = 0) const
{
    if (!*this) return;
    uint64_t rot_key{m_rotations[key_offset_bytes % SIZE_BYTES]}; // Continue obfuscation from where we left off
    const size_t alignment_remaining{std::min(SIZE_BYTES - (reinterpret_cast<ptrdiff_t>(target.data()) % SIZE_BYTES), target.size())};
    Xor(target, rot_key, alignment_remaining);
    target = target.subspan(alignment_remaining);
    rot_key = m_rotations[(key_offset_bytes + alignment_remaining) % SIZE_BYTES];
    for (; target.size() >= SIZE_BYTES; target = target.subspan(SIZE_BYTES)) { // Process multiple bytes at a time
        *reinterpret_cast<uint64_t*>(target.data()) ^= rot_key;
    }
    Xor(target, rot_key, target.size());
}
```

</details>

|              ns/MiB |               MiB/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           19,552.98 |           51,143.11 |    0.1% |     11.01 | `ObfuscationBench`
|           19,529.72 |           51,204.01 |    0.2% |     11.00 | `ObfuscationBench`
|           20,046.53 |           49,883.94 |    0.1% |     11.03 | `ObfuscationBench`


----

However, looking at the assembly closer, I reiterated why gcc doesn't do any SIMD and experimented with a few options and it seems we can help GCC by doing a big unrolled iteration before the whole thing:
```
for (constexpr auto unroll{8}; target.size() >= SIZE_BYTES * unroll; target = target.subspan(SIZE_BYTES * unroll)) {
    for (size_t i{0}; i < unroll; ++i) {
        Xor(target.subspan(i * SIZE_BYTES, SIZE_BYTES), rot_key, SIZE_BYTES);
    }
}
```
which helps GCC recognise that the loop is free of dependencies and performs a 128-bit SIMD pxor over 64-byte blocks instead of doing two 64-bit scalar xors per 16-byte stride - speeding up the benchmark from the mentioned `~19 GiB/s` to `~27 GiB/s` - i.e. additional 40% speedup. (Clang is unaffected)
I'll do an IBD again to see if it's significant enough to include it here.

Reply:
> do you see an explanation in the assembly why it would be faster?

Didn't check the assembly yet, do you mind sharing your workflow for that? I was using a combination of `objdump` with a bunch of flags + searching in `less` for the consteval hex stuff ~9 months ago but suspect `Obfuscation`-methods get heavily inlined?

Re-measured with Clang to make sure I wasn't running the wrong build, but can still reproduce the result:

At tip of PR 989537ff4026d7c3fa5ba99701e0a4b134d950f7:

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           73,504.93 |           13,604.53 |    0.1% |      655,366.68 |      270,831.04 |  2.420 |     131,074.07 |    0.0% |     10.66 | `ObfuscationBench`

With diff from https://github.com/bitcoin/bitcoin/pull/31144#discussion_r2097271922 applied:

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           37,501.52 |           26,665.59 |    0.9% |      294,930.14 |      138,093.33 |  2.136 |      32,771.94 |    0.0% |     11.03 | `ObfuscationBench`

> `xor_bytes_reference` randomly generates where the next offset will be, check it by adding:

That only tests with an offset into the key, not an offset into the target, right?

I'm thinking something like:
```diff
diff --git a/src/test/streams_tests.cpp b/src/test/streams_tests.cpp
index 1985b24663..2e10792710 100644
--- a/src/test/streams_tests.cpp
+++ b/src/test/streams_tests.cpp
@@ -61,6 +61,7 @@ BOOST_AUTO_TEST_CASE(xor_bytes_reference)
     for (size_t test{0}; test < 100; ++test) {
         const size_t write_size{1 + m_rng.randrange(100U)};
         const size_t key_offset{m_rng.randrange(3 * 8U)}; // Should wrap around
+        const size_t write_offset{std::min(write_size, m_rng.randrange(Obfuscation::SIZE_BYTES * 2))};
 
         std::array<std::byte, Obfuscation::SIZE_BYTES> key_bytes{};
         if (m_rng.randbool()) m_rng.fillrand(key_bytes);
@@ -68,8 +69,8 @@ BOOST_AUTO_TEST_CASE(xor_bytes_reference)
         std::vector expected{m_rng.randbytes<std::byte>(write_size)};
         std::vector actual{expected};
 
-        expected_xor(expected, key_bytes, key_offset);
-        obfuscation(actual, key_offset);
+        expected_xor(std::span{expected}.subspan(write_offset), key_bytes, key_offset);
+        obfuscation(std::span{actual}.subspan(write_offset), key_offset);
 
         BOOST_CHECK_EQUAL_COLLECTIONS(expected.begin(), expected.end(), actual.begin(), actual.end());
     }
```
It's super unlikely that one would start at an unaligned offset like that, but still more complete to test it.

Reply:
Wasn't able to use the small Godbolt sample to divine the reason for my 2x Clang speedup yet.

Fell back to `objdump -gSC build/bin/bench_bitcoin` and clearly see:
```asm
            *reinterpret_cast<uint64_t*>(target.data()) ^= rot_key;
  1107c0:       f3 0f 6f 4c d1 f0       movdqu -0x10(%rcx,%rdx,8),%xmm1
  1107c6:       f3 0f 6f 14 d1          movdqu (%rcx,%rdx,8),%xmm2
  1107cb:       66 0f ef c8             pxor   %xmm0,%xmm1
  1107cf:       66 0f ef d0             pxor   %xmm0,%xmm2
  1107d3:       f3 0f 7f 4c d1 f0       movdqu %xmm1,-0x10(%rcx,%rdx,8)
  1107d9:       f3 0f 7f 14 d1          movdqu %xmm2,(%rcx,%rdx,8)
  1107de:       48 83 c2 04             add    $0x4,%rdx
  1107e2:       48 39 d0                cmp    %rdx,%rax
  1107e5:       75 d9                   jne    1107c0 <Obfuscation::operator()(std::span<std::byte, 18446744073709551615ul>, unsigned long) const+0xf0>
```
from my suggestion, close to the `ObfuscationBench` symbols. But I couldn't spot `pxor`/`xmm` close-by in the baseline dump.

Maybe too synthetic to bother with.

Reply:
Took me longer than anticipated, but I have measured different scenarios here to understand where alignment matters and where it doesn't, with gcc, clang and apple clang

GCC 14 is 38.5% faster with alignment & 64 byte unroll compared to just a 64 bit xor

<details>
<summary>------ C++ compiler .......................... GNU 14.2.0 ------</summary>

```bash
rm -rf build && cmake -B build -DBUILD_BENCH=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ >/dev/null 2>&1 && cmake --build build -j$(nproc) >/dev/null 2>&1\
&& build/bin/bench_bitcoin -filter='ObfuscationBench' -min-time=5000 \
&& build/bin/bench_bitcoin -filter='ObfuscationBench' -min-time=5000 \
&& git log -1 --pretty=format:'%h %s'
```

> 00d77f4097 refactor: unify xor vs obfuscation nomenclature

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          879,957.05 |            1,136.42 |    0.0% |    9,437,186.42 |    3,158,477.41 |  2.988 |   1,048,576.99 |    0.0% |      5.50 | `ObfuscationBench`
|          879,916.05 |            1,136.47 |    0.0% |    9,437,186.42 |    3,157,995.25 |  2.988 |   1,048,576.99 |    0.0% |      5.50 | `ObfuscationBench`


> 2daaa83070 optimization: migrate fixed-size obfuscation from `std::vector<std::byte>` to `uint64_t`

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           51,665.69 |           19,355.21 |    0.0% |      327,684.05 |      185,409.22 |  1.767 |      65,536.55 |    0.0% |      5.50 | `ObfuscationBench`
|           52,520.96 |           19,040.02 |    0.2% |      327,684.05 |      188,497.60 |  1.738 |      65,536.55 |    0.0% |      5.51 | `ObfuscationBench`

> c69f7440db experiment: aligned xor only

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           51,601.83 |           19,379.16 |    0.0% |      327,687.85 |      185,151.71 |  1.770 |      65,536.95 |    0.0% |      5.50 | `ObfuscationBench`
|           51,952.63 |           19,248.30 |    0.1% |      327,687.85 |      186,403.90 |  1.758 |      65,536.95 |    0.0% |      5.50 | `ObfuscationBench`

> 9472af65af experiment: unroll only

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           46,341.13 |           21,579.10 |    0.0% |      475,139.45 |      166,272.12 |  2.858 |      81,920.35 |    0.0% |      5.50 | `ObfuscationBench`
|           45,936.99 |           21,768.95 |    0.0% |      475,139.45 |      164,853.63 |  2.882 |      81,920.35 |    0.0% |      5.50 | `ObfuscationBench`

> e9fd88bb3b experiment: unroll + alignment

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           37,281.01 |           26,823.31 |    0.1% |      475,139.64 |      133,808.24 |  3.551 |      81,920.54 |    0.0% |      5.35 | `ObfuscationBench`
|           37,652.77 |           26,558.47 |    0.1% |      475,139.64 |      135,132.04 |  3.516 |      81,920.54 |    0.0% |      5.35 | `ObfuscationBench`

</details>


LLVM Clang20 is 2.15x faster with alignment & 64 byte unroll compared to just a 64 bit xor
<details>
<summary>------ C++ compiler .......................... Clang 20.1.6 ------</summary>

```bash
rm -rf build && cmake -B build -DBUILD_BENCH=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ >/dev/null 2>&1 && cmake --build build -j$(nproc) >/dev/null 2>&1 \
&& build/bin/bench_bitcoin -filter='ObfuscationBench' -min-time=5000 \
&& build/bin/bench_bitcoin -filter='ObfuscationBench' -min-time=5000 \
&& git log -1 --pretty=format:'%h %s'
```

> 00d77f4097 refactor: unify xor vs obfuscation nomenclature

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          926,379.31 |            1,079.47 |    0.1% |    6,815,747.36 |    3,325,871.40 |  2.049 |     524,289.23 |    0.0% |      5.50 | `ObfuscationBench`
|          926,608.63 |            1,079.20 |    0.0% |    6,815,747.36 |    3,326,572.88 |  2.049 |     524,289.23 |    0.0% |      5.50 | `ObfuscationBench`

> 2daaa83070 optimization: migrate fixed-size obfuscation from `std::vector<std::byte>` to `uint64_t`

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           74,817.84 |           13,365.80 |    0.0% |      655,366.68 |      268,566.88 |  2.440 |     131,074.08 |    0.0% |      5.50 | `ObfuscationBench`
|           75,351.98 |           13,271.05 |    0.1% |      655,366.68 |      270,405.84 |  2.424 |     131,074.08 |    0.0% |      5.51 | `ObfuscationBench`

> c69f7440db experiment: aligned xor only

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           38,396.57 |           26,043.99 |    0.1% |      294,929.64 |      137,732.80 |  2.141 |      32,771.94 |    0.0% |      5.35 | `ObfuscationBench`
|           38,151.47 |           26,211.31 |    0.2% |      294,929.64 |      136,877.12 |  2.155 |      32,771.94 |    0.0% |      5.38 | `ObfuscationBench`

> 9472af65af experiment: unroll only

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           35,556.53 |           28,124.23 |    0.1% |      262,148.44 |      127,610.00 |  2.054 |      16,384.64 |    0.0% |      5.32 | `ObfuscationBench`
|           36,409.31 |           27,465.50 |    0.1% |      262,148.44 |      130,650.29 |  2.006 |      16,384.64 |    0.0% |      5.33 | `ObfuscationBench`

> e9fd88bb3b experiment: unroll + alignment

|              ns/MiB |               MiB/s |    err% |         ins/MiB |         cyc/MiB |    IPC |        bra/MiB |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|           34,770.15 |           28,760.30 |    0.0% |      262,149.44 |      124,756.60 |  2.101 |      16,384.83 |    0.0% |      5.32 | `ObfuscationBench`
|           36,214.90 |           27,612.94 |    0.0% |      262,149.44 |      129,932.72 |  2.018 |      16,384.84 |    0.0% |      5.33 | `ObfuscationBench`

</details>

Combining unrolling with aligning readjusts the speed for Apple Clang (which was already doing these automatically so there's no speed difference because of the latest commit) and brings Clang and GCC closer to the max bus speed.

Clang needed your alignment to achieve that, and GCC needed the unrolled 64-byte XOR to take advantage of SIMD.
Added this as a separate commit with you as coauthor, explaining the extra optimizations + updated godbolt and table from

Target & Compiler | Stride (per iter) | Main operation(s) in loop | Effective XORs
-- | -- | -- | --
Clang x86-64 (trunk) | 32 bytes | two unaligned 128-bit loads → pxor with broadcast key → stores | 4 × 64-bit
GCC x86-64 (trunk) | 16 bytes | two scalar 64-bit xor reg-mem instructions directly on the target | 2 × 64-bit
GCC RV32 (trunk) | 8 bytes | copy 8 B to stack scratch → XOR via two 32-bit regs → copy back | 1 × 64-bit
GCC s390x (big-endian 14.2) | 64 bytes | eight XC ops using key cached on stack | 8 × 64-bit

to:

Target & Compiler | Stride (per hot-loop iter) | Main operation(s) in loop | Effective XORs / iter
-- | -- | -- | --
Clang x86-64 (trunk) | 64 bytes | 4 × movdqu → pxor → store (aligned after peel) | 8 × 64-bit
GCC x86-64 (trunk) | 64 bytes | same 4 × movdqu/pxor sequence, enabled by 8-way unroll | 8 × 64-bit
GCC RV32 (trunk) | 8 bytes | copy 8 B to temp stack → 2 × 32-bit XOR → copy back | 1 × 64-bit (as 2 × 32-bit)
GCC s390x (big-endian 14.2) | 64 bytes | 8 × XC (mem-mem 8-B XOR) with key cached on stack | 8 × 64-bit